keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -697,43 +705,58 @@ generatedFrom:
   apiGroup: dra.example.com
   uid: foobar-uid
 
-vendorParameters:
-  # A vendor can put any kind of object here to pass the configuration
-  # parameters down to the kubelet plugin. In this case, the vendor
-  # driver simply copied the entire CR. It could also be some
-  # separate, smaller configuration type.
+# Setup parameters can be provided at two different levels:
+# - for all resources (here)
+# - for a single resource (not shown below)
+#
+# At the moment, only setup parameters defined by a vendor
+# are supported. In-tree definition of common setup parameters
+# might get added in the future.
+setup:
+  vendor:
+  # The driver name specifies which driver the list entry is intended
+  # for. This is a list because a claim might end up using resources
+  # from different drivers.
   #
-  # Beware that ResourceClaimParameters have separate RBAC rules than
-  # the vendor CRD, so information included here may get visible
-  # to more users than the original CRD. Both objects are in the same
-  # namespace.
-  kind: CardParameters
-  apiVersion: dra.example.com/v1alpha1
-  metadata:
-    name: my-parameters
-    namespace: user-namespace
-    uid: foobar-uid
-  ...
-  spec:
-    minimumRuntimeVersion: v12.0.0
-    minimumMemory: 32Gi
-    sharing:
-      strategy: TimeSliced
+  # A driver could provide an admission webhook to validate its
+  # own parameters when users embed them in a ResourceClaimParameters
+  # object themselves.
+  - driverName: cards.dra.example.com
+    parameters:
+      # Beware that ResourceClaimParameters have separate RBAC rules than
+      # the vendor CRD, so information included here may get visible
+      # to more users than the original CRD. Both objects are in the same
+      # namespace.
+      kind: CardClaimSetup
+      apiVersion: dra.example.com/v1alpha1
+      sharing:
+        strategy: TimeSliced
 
-driverRequests:
-- driverName: cards.dra.example.com
-  requests:
-  # Each entry here is a request for one resource.
-  - namedresources:
-      # Selectors are CEL expressions with access to the attributes of the named resource
-      # that is being checked for a match.
-      selector: |-
-        attributes.version["runtimeVersion"].isGreaterThan(semver("12.0.0")) &&
-        attributes.quantity["memory"].isGreaterThan(quantity("32Gi"))
-```
+namedResources:
+- attributeSuffix: cards.dra.example.com
```

Comment:
I don't quite follow this. Does this mean that when i define my resource slice that I have to name my attributes with a particular suffix to match here? Or is this the driver name appended to the end of the attributes I define in my resourceslice?

Reply:
Attribute names get expanded in two places:
- In ResourceSlice, the driver name gets appended.
- In a selector, either the user already uses the fully-qualified name or the "attributeSuffix" gets appended.

I am open for dropping the "attributeSuffix" field. It helps with keeping CEL expressions shorter, but it also makes the API harder to understand.

Reply:
>  I definitely don't want to be that verbose when generating a resource slice.

So we keep what I have for ResourceSlice now, with short names that inherit the driver name as suffix.

> It also feels weird (at least to me) to encode the driver name in the attribute string at all

The problem is that we want to enable selectors that work across different vendors. The attributes then could come from a `k8s.io` scope (namespace?) or from the scopes of two different vendors, if that is what the user wants: "select a card of vendor A with these attributes or a card of vendor B with these attributes".

> Would an alternative be to have an optional field for driver at the same level as name/<attribute-type> in cases where you want to match on exactly that driver?

Field where? The CEL expression is a single string.

Reply:
I think I get what you meant. We can rename "attributeSuffix" to "driverName" and change the meaning to "first filter by that name, then check the CEL expression". Should be more efficient, too.

Update pushed.